### DISCUSSION TITLE
Berry/Zigbee reading attributes_final doesn't answer as described in docs

### DISCUSSION THREAD

**Discussion by suppeee** [arendst/Tasmota#22623] [Q&A] [answered]
### PROBLEM DESCRIPTION
_A clear and concise description of what the problem is._
The tasmota berry docs say, to "def attributes_final(event_type, frame, attr_list, idx)" it should print as example "shortaddr=OxC1BC event_type=attributes_final attr_list={"ActivePower":261.612,"(ActivePower)":"0B04/050B","Endpoint":21,"LinkQuality":21}"

With tasmota 14.3.0 on my ZB-GW03 it prints "shortaddr=Ox1C69 event_type=attributes_final attr_list={"AppVersion":101,"0000/FFE2":53,"0000/FFE4":0,"Endpoint":true,"LinkQuality":true}". 
I need the correct endpoint-number for further processing, not just "true".

The endpoint is correct on the level "event_type=frame_received", but starting with "event_type=attributes_raw" the endpoint switches to "true".

### REQUESTED INFORMATION
_Make sure your have performed every step and checked the applicable boxes before submitting your issue. Thank you!_

- [x ] Read the [Contributing Guide and Policy](https://github.com/arendst/Tasmota/blob/development/CONTRIBUTING.md) and [the Code of Conduct](https://github.com/arendst/Tasmota/blob/development/CODE_OF_CONDUCT.md)
- [x ] Searched the problem in [issues](https://github.com/arendst/Tasmota/issues)
- [x ] Searched the problem in [discussions](https://github.com/arendst/Tasmota/discussions)
- [x ] Searched the problem in the [docs](https://tasmota.github.io/docs/FAQ)
- [ ] Searched the problem in the [chat](https://discord.gg/Ks2Kzd4)
- [x ] Device used (e.g., Sonoff Basic): _____ ZB-GW03
- [ ] Tasmota binary firmware version number used: _____14.3.0
  - [ ] Pre-compiled
  - [x ] Self-compiled
- [x ] Flashing tools used: _____ Flash via web
- [ ] Provide the output of command: `Backlog Template; Module; GPIO 255`:
04:45:40.713 RSL: RESULT = {"NAME":"ZB-GW03-V1.2","GPIO":[0,0,3552,0,3584,0,0,0,5793,5792,320,544,5536,0,5600,0,0,0,0,5568,0,0,0,0,0,0,0,0,608,640,32,0,0,0,0,0],"FLAG":0,"BASE":1}
04:45:40.943 RSL: RESULT = {"Module":{"0":"ZB-GW03-V1.2"}}
04:45:41.200 RSL: RESULT = {"GPIO0":{"0":"None"},"GPIO1":{"0":"None"},"GPIO2":{"3552":"Zigbee Tx"},"GPIO3":{"0":"None"},"GPIO4":{"3584":"Zigbee Rx"},"GPIO5":{"0":"None"},"GPIO6":{"0":"None"},"GPIO7":{"0":"None"},"GPIO8":{"0":"None"},"GPIO9":{"0":"None"},"GPIO10":{"0":"None"},"GPIO11":{"0":"None"},"GPIO12":{"5793":"Zigbee Rst2"},"GPIO13":{"5792":"Zigbee Rst1"},"GPIO14":{"320":"Led_i1"},"GPIO15":{"544":"LedLink"},"GPIO16":{"5536":"ETH POWER"},"GPIO17":{"0":"None"},"GPIO18":{"5600":"ETH MDIO"},"GPIO19":{"0":"None"},"GPIO20":{"0":"None"},"GPIO21":{"0":"None"},"GPIO22":{"0":"None"},"GPIO23":{"5568":"ETH MDC"},"GPIO24":{"0":"None"},"GPIO25":{"0":"None"},"GPIO26":{"0":"None"},"GPIO27":{"0":"None"},"GPIO32":{"608":"I2C SCL1"},"GPIO33":{"640":"I2C SDA1"},"GPIO34":{"32":"Button1"},"GPIO35":{"0":"None"},"GPIO36":{"0":"None"},"GPIO37":{"0":"None"},"GPIO38":{"0":"None"},"GPIO39":{"0":"None"}}
- [ ] 
```lua
  Configuration output here:


```
- [ ] If using rules, provide the output of this command: `Backlog Rule1; Rule2; Rule3`:
```lua
  Rules output here:


```
- [ ] Provide the output of this command: `Status 0`:
```lua
  STATUS 0 output here:
04:46:45.165 RSL: STATUS = {"Status":{"Module":0,"DeviceName":"Tasmota","FriendlyName":["Tasmota"],"Topic":"tasmota_C25A74","ButtonTopic":"0","Power":"0","PowerLock":"0","PowerOnState":3,"LedState":1,"LedMask":"FFFF","SaveData":1,"SaveState":1,"SwitchTopic":"0","SwitchMode":[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],"ButtonRetain":0,"SwitchRetain":0,"SensorRetain":0,"PowerRetain":0,"InfoRetain":0,"StateRetain":0,"StatusRetain":0}}
04:46:45.169 RSL: STATUS1 = {"StatusPRM":{"Baudrate":115200,"SerialConfig":"8N1","GroupTopic":"tasmotas","OtaUrl":"","RestartReason":"Software reset CPU","Uptime":"0T22:56:21","StartupUTC":"2024-12-09T21:50:24","Sleep":50,"CfgHolder":4617,"BootCount":52,"BCResetTime":"2023-10-19T09:41:15","SaveCount":107}}
04:46:45.172 RSL: STATUS2 = {"StatusFWR":{"Version":"14.3.0(tasmota)","BuildDateTime":"2024-12-09T21:41:22","Core":"3_1_0","SDK":"5.3.1.240924","CpuFrequency":240,"Hardware":"ESP32-D0WD-V3 v3.1","CR":"314/699"}}
04:46:45.178 RSL: STATUS3 = {"StatusLOG":{"SerialLog":2,"WebLog":2,"MqttLog":0,"SysLog":3,"LogHost":"","LogPort":514,"SSId":["",""],"TelePeriod":300,"Resolution":"55C180C0","SetOption":["00000001","2805C80001000600003C5A0A192800000000","000000A0","00004000","00004000","00000000"]}}
04:46:45.186 RSL: STATUS4 = {"StatusMEM":{"ProgramSize":1531,"Free":1349,"Heap":144,"StackLowMark":3,"PsrMax":0,"PsrFree":0,"ProgramFlashSize":4096,"FlashSize":4096,"FlashChipId":"1640EF","FlashFrequency":40,"FlashMode":"DIO","Features":["0809","9F8A07DF","00500001","000000CF","010013C0","00000000","40024080","00200000","0400002C","00000080","00000004"],"Drivers":"1,!2,!3,!4,!5,!8,9,10,11,!12,20,!21,23,38,!41,50,52,62,!68,82,121","Sensors":"1,2,3,5,6,127","I2CDriver":"7"}}
04:46:45.191 RSL: STATUS5 = {"StatusNET":{"Hostname":"tasmota-C25A74-6772","IPAddress":"0.0.0.0","Gateway":"192.168.178.1","Subnetmask":"255.255.255.0","DNSServer1":"192.168.178.1","DNSServer2":"2a02:2761:21f8:1700:464e:6dff:fe33:8b35","Mac":"24:DC:C3:C2:5A:74","IP6Global":"","IP6Local":"","Ethernet":{"Hostname":"tasmota-C25A74-6772-eth","IPAddress":"192.168.178.30","Gateway":"192.168.178.1","Subnetmask":"255.255.255.0","DNSServer1":"192.168.178.1","DNSServer2":"2a02:2761:21f8:1700:464e:6dff:fe33:8b35","Mac":"24:DC:C3:C2:5A:77","IP6Global":"2a02:2761:21f8:1700:26dc:c3ff:fec2:5a77","IP6Local":"fe80::26dc:c3ff:fec2:5a77%en1"},"Webserver":2,"HTTP_API":1,"WifiConfig":4,"WifiPower":17.0}}
04:46:45.197 RSL: STATUS7 = {"StatusTIM":{"UTC":"2024-12-10T20:46:45Z","Local":"2024-12-11T04:46:45","StartDST":"2024-03-31T02:00:00","EndDST":"2024-10-27T03:00:00","Timezone":"+08:00","Sunrise":"15:34","Sunset":"23:52"}}
04:46:45.200 RSL: STATUS10 = {"StatusSNS":{"Time":"2024-12-11T04:46:45"}}
04:46:45.203 RSL: STATUS11 = {"StatusSTS":{"Time":"2024-12-11T04:46:45","Uptime":"0T22:56:21","UptimeSec":82581,"Heap":143,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":0,"Berry":{"HeapUsed":5,"Objects":88}}}

```
- [ ] Set `weblog` to 4 and then, when you experience your issue, provide the output of the Console log:
```lua
  Console output here:


```

### TO REPRODUCE
_Steps to reproduce the behavior:_

to "def attributes_final(event_type, frame, attr_list, idx)" my ZB-GW03 prints "shortaddr=Ox1C69 event_type=attributes_final attr_list={"AppVersion":101,"0000/FFE2":53,"0000/FFE4":0,"Endpoint":true,"LinkQuality":true}". 

### EXPECTED BEHAVIOUR
_A clear and concise description of what you expected to happen._

The tasmota berry docs say, to "def attributes_final(event_type, frame, attr_list, idx)" it should print as example "shortaddr=OxC1BC event_type=attributes_final attr_list={"ActivePower":261.612,"(ActivePower)":"0B04/050B","Endpoint":21,"LinkQuality":21}"

I need the correct endpoint-number for further processing, not just "true".
The endpoint is correct on the level "event_type=frame_received", but starting with "event_type=attributes_raw" the endpoint switches to "true".

### SCREENSHOTS
_If applicable, add screenshots to help explain your problem._

### ADDITIONAL CONTEXT
_Add any other context about the problem here._

**(Please, remember to close the issue when the problem has been addressed)**

**Comment by sfromis** [comment]
Hard to reproduce without even knowing what the Zigbee device is, and without the log at the relevant time. It might depend on what device it is, and details from the log.

With one of my devices I'm seeing: `"Endpoint":1,"LinkQuality":185`
This indicates that Tasmota can work as you expected, but something in your setup differs.

**Reply by suppeee** [reply]
Thanks for your reply and for bothering with my problem.

I use several zigbee devices. The log with the berry code example for zigbee is on the console as follows. As you can see, the RSL: SENSOR lines show the Endpoint, however on the level event_type=attributes_final there is just an "true".

i compiled tasmota 14.3.0 with the instructions on https://github.com/xsp1989/ZB-GW03_firmware/blob/master/tasmota/README.md and gitpod
i saw, tasmota 14.4 was released today I will compile that tomorrow and try again.

if you need further logs for investigating, i would be more than happy to provide them.

Aqara Cube (https://zigbee.blakadder.com/Aqara_MFKZQ01LM.html) 
```
05:27:10.282 shortaddr=Ox6A66 event_type=frame_received frame={'srcendpoint': 2, 'transactseq_set': 0, 'shortaddr': 27238, 'dstendpoint': 1, 'payload': bytes('5500218300'), 'shortaddr_hex': '0x6A66', 'manuf': 0, 'payload_ptr': <ptr: 0x3ffd8c68>, 'need_response': 0, 'transactseq': 172, 'cmd': 10, 'direct': 0, 'cluster': 18, 'cluster_specific': 0, 'groupaddr': 0}
05:27:10.294 shortaddr=Ox6A66 event_type=attributes_raw attr_list={"0012/0055":131,"Endpoint":true,"LinkQuality":true}
05:27:10.335 shortaddr=Ox6A66 event_type=attributes_refined attr_list={"AqaraCube":"flip180","AqaraCubeSide":3,"Endpoint":true,"LinkQuality":true}
05:27:10.717 shortaddr=Ox6A66 event_type=attributes_final attr_list={"AqaraCube":"flip180","AqaraCubeSide":3,"Endpoint":true,"LinkQuality":true}
05:27:10.720 RSL: SENSOR = {"ZbReceived":{"0x6A66":{"Device":"0x6A66","AqaraCube":"flip180","AqaraCubeSide":3,"Endpoint":2,"LinkQuality":97}}}
```

Aqara D1 wireless switch (https://zigbee.blakadder.com/Aqara_WXKG07LM.html)
```
05:01:01.216 shortaddr=OxD6AA event_type=frame_received frame={'srcendpoint': 2, 'transactseq_set': 0, 'shortaddr': 54954, 'dstendpoint': 1, 'payload': bytes('5500210100'), 'shortaddr_hex': '0xD6AA', 'manuf': 0, 'payload_ptr': <ptr: 0x3ffb40fc>, 'need_response': 0, 'transactseq': 158, 'cmd': 10, 'direct': 0, 'cluster': 18, 'cluster_specific': 0, 'groupaddr': 0}
05:01:01.226 shortaddr=OxD6AA event_type=attributes_raw attr_list={"0012/0055":1,"Endpoint":true,"LinkQuality":true}
05:01:01.282 shortaddr=OxD6AA event_type=attributes_refined attr_list={"MultiInValue":1,"(MultiInValue)":"0012/0055","Click":"single","click":"single","Endpoint":true,"LinkQuality":true}
05:01:01.668 shortaddr=OxD6AA event_type=attributes_final attr_list={"MultiInValue":1,"Click":"single","click":"single","Endpoint":true,"LinkQuality":true}
05:01:01.671 RSL: SENSOR = {"ZbReceived":{"0xD6AA":{"Device":"0xD6AA","Name":"Taster","MultiInValue":1,"Click":"single","click":"single","Endpoint":2,"LinkQuality":21}}}
05:01:02.776 shortaddr=OxD6AA event_type=frame_received frame={'srcendpoint': 1, 'transactseq_set': 0, 'shortaddr': 54954, 'dstendpoint': 1, 'payload': bytes('5500210100'), 'shortaddr_hex': '0xD6AA', 'manuf': 0, 'payload_ptr': <ptr: 0x3ffd26f0>, 'need_response': 0, 'transactseq': 159, 'cmd': 10, 'direct': 0, 'cluster': 18, 'cluster_specific': 0, 'groupaddr': 0}
05:01:02.787 shortaddr=OxD6AA event_type=attributes_raw attr_list={"0012/0055":1,"Endpoint":true,"LinkQuality":true}
05:01:02.842 shortaddr=OxD6AA event_type=attributes_refined attr_list={"MultiInValue":1,"(MultiInValue)":"0012/0055","Click":"single","click":"single","Endpoint":true,"LinkQuality":true}
05:01:03.227 shortaddr=OxD6AA event_type=attributes_final attr_list={"MultiInValue":1,"Click":"single","click":"single","Endpoint":true,"LinkQuality":true}
05:01:03.230 RSL: SENSOR = {"ZbReceived":{"0xD6AA":{"Device":"0xD6AA","Name":"Taster","MultiInValue":1,"Click":"single","click":"single","Endpoint":1,"LinkQuality":21}}}
```

LEDVANCE A60 RGBW Value II:
```
05:02:01.151 shortaddr=Ox9EB2 event_type=frame_received frame={'srcendpoint': 1, 'transactseq_set': 0, 'shortaddr': 40626, 'dstendpoint': 1, 'payload': bytes('00002074010020FE03002101000400210100070021990008003000'), 'shortaddr_hex': '0x9EB2', 'manuf': 0, 'payload_ptr': <ptr: 0x3ffd8bb4>, 'need_response': 1, 'transactseq': 20, 'cmd': 10, 'direct': 0, 'cluster': 768, 'cluster_specific': 0, 'groupaddr': 0}
05:02:01.182 shortaddr=Ox9EB2 event_type=attributes_raw attr_list={"0300/0000":116,"0300/0001":254,"0300/0003":1,"0300/0004":1,"0300/0007":153,"0300/0008":0,"Endpoint":true,"LinkQuality":true}
05:02:01.322 shortaddr=Ox9EB2 event_type=attributes_refined attr_list={"Hue":116,"(Hue)":"0300/0000","Sat":254,"(Sat)":"0300/0001","X":1,"(X)":"0300/0003","Y":1,"(Y)":"0300/0004","CT":153,"(CT)":"0300/0007","ColorMode":0,"(ColorMode)":"0300/0008","RGB":"00FFBB","RGBb":"00FCB9","Endpoint":true,"LinkQuality":true}
05:02:01.731 shortaddr=Ox9EB2 event_type=attributes_final attr_list={"Power":1,"Dimmer":251,"Hue":116,"Sat":254,"X":1,"Y":1,"CT":153,"ColorMode":0,"RGB":"00FFBB","RGBb":"00FCB9","Endpoint":true,"LinkQuality":true}
05:02:01.734 RSL: SENSOR = {"ZbReceived":{"0x9EB2":{"Device":"0x9EB2","Power":1,"Dimmer":251,"Hue":116,"Sat":254,"X":1,"Y":1,"CT":153,"ColorMode":0,"RGB":"00FFBB","RGBb":"00FCB9","Endpoint":1,"LinkQuality":8}}}
```

Dimmer TS0501B - _TZ3210_agjx0pxt_
```
5:11:55.237 shortaddr=Ox1C69 event_type=frame_received frame={'srcendpoint': 1, 'transactseq_set': 0, 'shortaddr': 7273, 'dstendpoint': 1, 'payload': bytes('01002065E2FF2035E4FF2000'), 'shortaddr_hex': '0x1C69', 'manuf': 0, 'payload_ptr': <ptr: 0x3ffb40dc>, 'need_response': 1, 'transactseq': 112, 'cmd': 10, 'direct': 0, 'cluster': 0, 'cluster_specific': 0, 'groupaddr': 0}
05:11:55.257 shortaddr=Ox1C69 event_type=attributes_raw attr_list={"0000/0001":101,"0000/FFE2":53,"0000/FFE4":0,"Endpoint":true,"LinkQuality":true}
05:11:55.327 shortaddr=Ox1C69 event_type=attributes_refined attr_list={"AppVersion":101,"(AppVersion)":"0000/0001","0000/FFE2":53,"0000/FFE4":0,"Endpoint":true,"LinkQuality":true}
05:11:55.341 shortaddr=Ox1C69 event_type=attributes_final attr_list={"AppVersion":101,"0000/FFE2":53,"0000/FFE4":0,"Endpoint":true,"LinkQuality":true}
05:11:55.344 RSL: SENSOR = {"ZbReceived":{"0x1C69":{"Device":"0x1C69","Name":"Strip","AppVersion":101,"0000/FFE2":53,"0000/FFE4":0,"Endpoint":1,"LinkQuality":97}}}
```

**Reply by suppeee** [reply]
i'm sorry to say that the situation is unchanged with 14.4.1 compiled today.
and obviously the problem is not restricted to the endpoint number but also applies to the linkquality-value.

any ideas? 
retrieving the information from the event_type=frame_received and glueing it together is possible in principle but adds more complexity.
as alternative, can i access the information on console in the line RSL: SENSOR with berry?

**Comment by s-hadinger** [comment, marked as answer]
Thanks for reporting this bug. It is now fixed in the latest version in the `development` branch.

**Reply by suppeee** [reply]
Thank you very much!
